kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchConnectionManager.java

```diff
@@ -124,12 +128,26 @@ private final void runWatch() throws MalformedURLException, ExecutionException,
       private final Logger logger = LoggerFactory.getLogger(this.getClass());
 
       @Override
-      public void onOpen(WebSocket webSocket, Response response) {
+      public void onOpen(final WebSocket webSocket, Response response) {
         logger.debug("WebSocket successfully opened");
         webSocketRef.set(webSocket);
         currentReconnectAttempt.set(0);
         started.set(true);
+        queue.clear();
         queue.add(true);
+        pingFuture = pingExecutor.scheduleAtFixedRate(
+          new Runnable() {
+            @Override
+            public void run() {
+              try {
+                webSocket.sendPing(new Buffer().writeUtf8("Alive?"));
+              } catch (IOException e) {
+                logger.error("Failed to send ping", e);
+                onFailure(e, null);
```

Comment:
I always thought that the onFailure() was referring to the upgrade operation and was not expected after a successful upgrade. Still the javadoc doesn't verify my assumption. 

Need to be careful, if there is code written using this assumption.